zkevm-circuits/src/evm_circuit/execution/begin_tx.rs

```diff
@@ -137,15 +158,74 @@ impl<F: Field> ExecutionGadget<F> for BeginTxGadget<F> {
         );
 
         // Transfer value from caller to callee
+        let intrinsic_tx_value = cb.query_word();
+        let intrinsic_mul_gas_fee_by_gas = cb.query_word();
+        let intrinsic_tx_value_and_gas_fee_is_zero = BatchedIsZeroGadget::construct(
+            cb,
+            [
+                intrinsic_tx_value.expr(),
+                intrinsic_mul_gas_fee_by_gas.expr(),
+            ],
+        );
+        cb.condition(is_tx_invalid.expr(), |cb| {
+            cb.require_equal(
+                "intrinsic_tx_value == 0 && intrinsic_mul_gas_fee_by_gas == 0",
+                intrinsic_tx_value_and_gas_fee_is_zero.expr(),
+                true.expr(),
+            );
+        });
+
+        cb.condition(1.expr() - is_tx_invalid.expr(), |cb| {
+            cb.require_equal(
+                "intrinsic_tx_value == tx_value",
+                intrinsic_tx_value.expr(),
+                tx_value.expr(),
+            );
+            cb.require_equal(
+                "intrinsic_tx_value == tx_value",
+                intrinsic_mul_gas_fee_by_gas.expr(),
+                mul_gas_fee_by_gas.product().expr(),
+            );
+        });
+
         let transfer_with_gas_fee = TransferWithGasFeeGadget::construct(
             cb,
             tx_caller_address.expr(),
             tx_callee_address.expr(),
-            tx_value.clone(),
-            mul_gas_fee_by_gas.product().clone(),
+            intrinsic_tx_value.clone(),
+            intrinsic_mul_gas_fee_by_gas.clone(),
```

Comment:
Strange names because these don't have anything to do with the intrinsic gas cost? Something like `effective_tx_value` and `efftive_gas_fee` seems to make more sense to me.

Reply:
Updated to `effective_tx_value` and `effective_gas_fee`